client/session.go

```diff
@@ -1625,74 +1661,163 @@ func (s *session) FetchBlocksFromPeers(
 
 func (s *session) streamBlocksMetadataFromPeers(
 	namespace ident.ID,
-	shard uint32,
-	peers []peer,
+	shardID uint32,
+	peers peers,
 	start, end time.Time,
+	consistencyLevel ReadConsistencyLevel,
 	metadataCh chan<- blocksMetadata,
 	resultOpts result.Options,
 	progress *streamFromPeersMetrics,
 	version FetchBlocksMetadataEndpointVersion,
 ) error {
 	var (
 		wg       sync.WaitGroup
-		errLock  sync.Mutex
-		errLen   int
-		pending  int64
-		multiErr = xerrors.NewMultiError()
+		errLock  sync.RWMutex
+		errors   = make(map[int]error)
+		setError = func(idx int, err error) {
+			errLock.Lock()
+			errors[idx] = err
+			errLock.Unlock()
+		}
+		getErrors = func() []error {
+			var result []error
+			errLock.RLock()
+			for _, err := range errors {
+				if err == nil {
+					continue
+				}
+				result = append(result, err)
+			}
+			errLock.RUnlock()
+			return result
+		}
+		abortError    error
+		setAbortError = func(err error) {
+			errLock.Lock()
+			abortError = err
+			errLock.Unlock()
+		}
+		getAbortError = func() error {
+			errLock.RLock()
+			result := abortError
+			errLock.RUnlock()
+			return result
+		}
+		pending   = int64(len(peers.peers))
+		majority  = atomic.LoadInt32(&s.majority)
+		enqueued  = int32(len(peers.peers))
+		responded int32
+		success   int32
 	)
+	if peers.selfExcluded {
+		state, err := peers.selfHostShardSet.ShardSet().LookupStateByID(shardID)
+		if err == nil && state == shard.Available {
+			// If we excluded ourselves from fetching, we basically treat ourselves
```

Comment:
This comment seems a little misleading. We can't really bootstrap from ourselves because being down for any amount of time implies that we are missing data.

Maybe re-phrase to say that as long as we read metadata from quorum nodes (including ourselves) we're guaranteed to not miss any confirmed writes? Although that implies that we read our own commit log.

Reply:
It's been rephrased somewhat, let me know if you don't think it's descriptive enough.